tools/libSwiftSyntaxParser/libSwiftSyntaxParser.cpp

```diff
@@ -235,3 +317,47 @@ swiftparse_parse_string(swiftparse_parser_t c_parser, const char *source) {
 const char* swiftparse_syntax_structure_versioning_identifier(void) {
   return getSyntaxStructureVersioningIdentifier();
 }
+
+//===--------------------- C API for diagnostics -------------------------====//
+
+void
+swiftparse_parser_set_diagnostic_handler(swiftparse_parser_t c_parser,
+                                         swiftparse_diagnostic_handler_t hdl) {
+  SynParser *parser = static_cast<SynParser*>(c_parser);
+  parser->setDiagnosticHandler(hdl);
+}
+
+const char* swiftparse_diagnostic_get_message(swiftparser_diagnostic_t diag) {
+  return static_cast<DiagnosticDetail*>(diag)->Message;
+}
+
+unsigned swiftparse_diagnostic_get_fixit_count(swiftparser_diagnostic_t diag) {
+  return static_cast<DiagnosticDetail*>(diag)->AllFixits.size();
+}
+
+swiftparse_fixit_t
+swiftparse_diagnostic_get_fixit_at_index(swiftparser_diagnostic_t diag,
+                                         unsigned idx) {
+  auto allFixits = static_cast<DiagnosticDetail*>(diag)->AllFixits;
+  assert(allFixits.size() >= idx);
```

Comment:
`idx` must be less than `size()` it will be buffer overflow if it is equal to `size()`.